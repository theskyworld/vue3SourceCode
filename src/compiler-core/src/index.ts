// 导出该模块中的所有文件，供外部导入使用
export * from "./compile";