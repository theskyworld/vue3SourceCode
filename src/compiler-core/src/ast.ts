import { CREATE_ELEMENT_BLOCK, OPENBLOCK } from "./utils/transform/runtimeHelpers";
import { isEmpty } from "./utils/utils";


export const enum NodeTypes {
    INTERPOLATION,
    SIMPLE_EXPRESSION,
    ELEMENT,
    TEXT,
    ROOT,
    COMPOUND_EXPRESSION,
}

export function createVNodeCall(context, tag, props, children) {
  // 实现元素和联合类型中的"const { openBlock: _openBlock, createElementBlock: _createElementBlock } = Vue"部分
  context.helper(CREATE_ELEMENT_BLOCK);
  context.helper(OPENBLOCK);

  return {
    type: NodeTypes.ELEMENT,
    tag,
    props,
    children,
  };
}