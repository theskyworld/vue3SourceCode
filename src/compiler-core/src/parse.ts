import { NodeTypes } from "./ast";

const enum TagType{
    Start,
    End,
}

export function baseParse(content : string) {

    const context = createParserContext(content);
    return createRoot(parseChildren(context, []))


}

function parseChildren(context, ancestors) {
    // 使用ancestors收集所有的标签名
    // 用于判断是否存在缺少结束标签的标签或者获取某个指定标签

    const nodes: Array<any> = [];
   
    // 循环解析一个或多个children
    while (!isEnd(context, ancestors)) {
        let node;
         // 判断解析插值
        if (context.source.startsWith("{{")) {
            node = parseInterpolation(context);
            // 判断解析元素
        } else if (context.source[0] === "<" && /[a-z]/i.test(context.source[1])) {
            // console.log('parse element')
            node = parseElement(context, ancestors);
        }

        // 判断解析文本
        if (!node) {
            node = parseText(context);
        }
        nodes.push(node);
    }

    return nodes;
}

function isEnd(context, ancestors) {
    // 确定while循环结束的条件
    // 1.当遇到结束标签时
    const s = context.source;
    // if (parentTag && s.startsWith(`</${parentTag}>`)) {
    //     return true;
    // }

    // 当每次遇到结束标签时，判断ancestors中是否存在相应匹配的开始标签
    if (s.startsWith("</")) {
        // for (let i = 0; i < ancestors.length; i++) {
        for (let i = ancestors.length - 1; i >= 0; i--) {
            // 取出ancestors中每个标签的标签名
            const tag = ancestors[i].tag;
            // 截取出结束标签的标签名并进行对比
            // if (s.slice(2, 2 + tag.length) === tag) {
            //     return true;
            // }

            if (isStartsWithEndTagOpen(s, tag)) {
                return true;
            }
        }
    }

    // 2.当source处理完成时
    return !s;

}

function parseText(context) {

    // 解析三种类型的联合类型
    // 字符串的部分是指在插值类型的"{{"前的部分
    let endIndex = context.source.length;
    let endTokens = ["<", "{{"];
    
    for (let i = 0; i < endTokens.length; i++) {
        const index = context.source.indexOf(endTokens[i]);
        // 在context.source中找到了"{{"，说明存在插值
        // 字符串的部分是指在插值类型的"{{"或者元素类型的"<"（嵌套元素内）两者之中第一个前的部分
        if (index !== -1 && endIndex > index) {
            endIndex = index;
        }
    }
    
    


    // // 获取字符串内容
    // const content = context.source.slice(0, endIndex);

    // // 删除已经处理的内容
    // advanceBy(context, content.length);
    const content = parseTextData(context, endIndex);

    


    return {
        type: NodeTypes.TEXT,
        content,
    }
}

function parseTextData(context, length) {
     // 获取字符串内
    const content = context.source.slice(0, length);

    // 删除已经处理的内容
    advanceBy(context, length);

    return content;
}

function parseElement(context, ancestors) {
    // 解析<div>
    const element: any = parseTag(context, TagType.Start);
    
    ancestors.push(element); //将每个element收集进入ancestors

    // 解析三种类型的联合类型
    // 解析元素类型的时候，其下面可能存在多个或多层嵌套的children需要解析
    // 递归进行解析
    element.children = parseChildren(context, ancestors);

    ancestors.pop();

    // if (context.source.slice(2, 2 + element.tag.length) === element.tag) {
    if(isStartsWithEndTagOpen(context.source, element.tag)){
            // 解析</div>
        parseTag(context, TagType.End);
    } else {
        throw new Error(`缺少结束标签:${element.tag}`)
    }

    return element;
}

function isStartsWithEndTagOpen(source, tag) {
    return source.startsWith("</") &&
        source.slice(2, 2 + tag.length).toLowerCase() === tag.toLowerCase();

}
function parseTag(context, tagType : TagType) {
    // 使用正则解析出tag
    const match: any = /^<\/?([a-z]*)/i.exec(context.source);
    // console.log(match); // "<div"
    const tag = match[1];

    // 删除处理完成的代码
    // 删除"<div>"
    advanceBy(context, match[0].length);
    advanceBy(context, 1);

    if (tagType === TagType.End) return;
    return {
        type: NodeTypes.ELEMENT,
        tag,
    }
}

function parseInterpolation(context) {
    // 解析接收到的插值
    // 取出字符串例如"{{message}}"中"{{}}"之间的内容message的值
    // "{{message}}"
    const openDelimiter = "{{";
    const closeDelimiter = "}}";

    const closeIndex = context.source.indexOf(closeDelimiter, closeDelimiter.length);
    // context.source = context.source.slice(openDelimiter.length)
    advanceBy(context, openDelimiter.length);

    // console.log(context.source); // "message}}"
    const rawContentLength = closeIndex - openDelimiter.length;
    // const rawcontent = context.source.slice(0, rawContentLength);
    const rawcontent = parseTextData(context, rawContentLength)
    // 处理空格
    const content = rawcontent.trim();
    // console.log(content); // "message"
    // 删除插值字符串后的（"}}"后的）所有内容

    // context.source = context.source.slice(rawContentLength + closeDelimiter.length);
    // advanceBy(context, rawContentLength + closeDelimiter.length);
    advanceBy(context, closeDelimiter.length);

    // console.log(content);
    return {
        // type: "interpolation",
        type : NodeTypes.INTERPOLATION,
        content: {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content : content,
        }
            
        
        
    }
}

function createRoot(children) {
    return {
        children,
        type : NodeTypes.ROOT,
    }
}

function createParserContext(content : string) : any {
    return {
        source : content,
    }
}

function advanceBy(context : any, length : number) {
    context.source = context.source.slice(length);;
}