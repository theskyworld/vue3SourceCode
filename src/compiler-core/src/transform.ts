// 支持接收options参数，从函数外部来指定函数内部对数据的处理方式

import { NodeTypes } from "./ast";
import { TO_DISPLAY_STRING } from "./utils/transform/runtimeHelpers";

// 实现函数的可测试性，函数的插件体系
export function transform(root, options = {}) {
    const context = createTransformContext(root, options);
    // 遍历语法树，深度优先搜索算法
    traverseNode(root, context);

    createRootCodegen(root);

    root.helpers = [...context.helpers.keys()]
}

function createRootCodegen(root) {
    // 获取语法树中的入口节点（root节点的第一个子节点，并赋值给root.codegenNode）
    // 以后处理语法树都基于入口节点处理
    // root.codegenNode = root.children[0];

    const child = root.children[0];
    if (child.type === NodeTypes.ELEMENT) {
        root.codegenNode = child.codegenNode;
    } else {
        root.codegenNode = root.children[0];
    }
}

function createTransformContext(root : any, options : any) : any {
    const context = {
        root,
        nodeTransforms: options.nodeTransforms || [],
        helpers: new Map(),
        helper(key) {
            context.helpers.set(key, 1);
        }
    }

    return context;
}


function traverseNode(node, context) {

    // 变动点
    // console.log(node); //打印遍历到的节点
    const nodeTransforms = context.nodeTransforms;
    const exitFns : any = [];
    for (let i = 0; i < nodeTransforms.length; i++) {
        const transform = nodeTransforms[i];
         // 修改text节点中的内容
        // 通过传入的options来对节点进行特定的修改
        // transform(node, context);
        const onExit = transform(node, context);
        if (onExit) exitFns.push(onExit);
    }

    switch (node.type) {
        case NodeTypes.INTERPOLATION:
            context.helper(TO_DISPLAY_STRING);
            break;
        case NodeTypes.ROOT:
        case NodeTypes.ELEMENT:
            traverseChildren(node, context);
            break;
        default:
            break;
    }

    let i = exitFns.length;
    while (i--) {
        exitFns[i]();
    }
    // 将函数的变动点和稳定点分离
    // 稳定点
    // traverseChildren(node, context);
}

function traverseChildren(node, context) {
    const children = node.children;

    // if (children) {
        for (let i = 0; i < children.length; i++) {
            const node = children[i];
            traverseNode(node, context);
        }
    // }
}