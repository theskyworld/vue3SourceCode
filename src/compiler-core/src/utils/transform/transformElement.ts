import { createVNodeCall, NodeTypes } from "../../ast";
import { CREATE_ELEMENT_BLOCK, OPENBLOCK } from "./runtimeHelpers";


// 实现元素类型中的"const { openBlock: _openBlock, createElementBlock: _createElementBlock } = Vue"部分
export function transformElement(node, context) {
    // const { helper } = context;
    // if (node.type === NodeTypes.ELEMENT) {
    //     helper(OPENBLOCK);
    //     helper(CREATE_ELEMENT_BLOCK);
    // }

    if (node.type === NodeTypes.ELEMENT) {
        return () => {
            // tag
            const vnodeTag = `${node.tag}`;

            // props
            let vnodeProps;

            // children
            const children = node.children;
            let vnodeChildren = children[0];

            node.codegenNode = createVNodeCall(
                context,
                vnodeTag,
                vnodeProps,
                vnodeChildren
            );
        };
  }
}