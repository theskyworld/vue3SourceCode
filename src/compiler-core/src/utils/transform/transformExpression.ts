import { NodeTypes } from "../../ast";

export function transformExpression(node, context) {
    if (node.type === NodeTypes.INTERPOLATION) {
        node.content = processExpression(node.content, context);
    }
}
function processExpression(node, context) {
    node.content = `_ctx.${node.content}`;
    return node;
}