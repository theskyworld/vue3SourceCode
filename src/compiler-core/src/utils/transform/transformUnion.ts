import { NodeTypes } from "../../ast"
import { CREATE_ELEMENT_BLOCK, OPENBLOCK, TO_DISPLAY_STRING } from "./runtimeHelpers"


// 实现联合类型中的"const { toDisplayString: _toDisplayString, openBlock: _openBlock, createElementBlock: _createElementBlock } = Vue"
export function transformUnion(node, context) {
    const { helper } = context
    if (node.type === NodeTypes.ELEMENT) {
        helper(OPENBLOCK);
        helper(CREATE_ELEMENT_BLOCK);
        node.children.forEach((child) => {
            if (child.type === NodeTypes.TEXT) {
                helper(TO_DISPLAY_STRING);
            };
        })
        
    }
}