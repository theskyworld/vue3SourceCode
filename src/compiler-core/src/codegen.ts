import { NodeTypes } from "./ast";
import { isEmpty } from "./utils/utils";
import { CREATE_ELEMENT_BLOCK, helperMapName, OPENBLOCK, TO_DISPLAY_STRING } from "./utils/transform/runtimeHelpers";
import { isString } from "../../shared";

// export function generate(ast) {
//     // 进行以下返回内容字符串的拼接生成
//     let code = '';
//     code += "return ";
//     const functionName = "render";
//     const args = ["_ctx", "_cache"];
//     const argsSignature = args.join(", ");
//     // const node = ast.children[0];
//     const node = ast.codegenNode;

//     code += `function ${functionName}(${argsSignature}){`;
//     code += `return '${node.content}'`;
//     code += "}";

//     return {
//         code,
//     }
// }


// 处理文本类型，最终要生成的模板为
// return function render(_ctx, _cache, $props, $setup, $data, $options) {
//   return "hi"
// }



// 处理插值类型，最终要生成的模板为
// const { toDisplayString: _toDisplayString } = Vue

// return function render(_ctx, _cache, $props, $setup, $data, $options) {
//   return _toDisplayString(_ctx.message)
// }


// 处理元素类型（<div></div>），最终要生成的模板为
// const { openBlock: _openBlock, createElementBlock: _createElementBlock } = Vue

// return function render(_ctx, _cache, $props, $setup, $data, $options) {
//   return (_openBlock(), _createElementBlock("div"))
// }


// 处理联合类型（<div>hi{{message}}</div>），最终要生成的模板为
// const { toDisplayString: _toDisplayString, openBlock: _openBlock, createElementBlock: _createElementBlock } = Vue

// return function render(_ctx, _cache, $props, $setup, $data, $options) {
//   return (_openBlock(), _createElementBlock("div", null, "hi" + _toDisplayString(_ctx.message), 1 /* TEXT */))
// }


// 逻辑抽离
export function generate(ast) {
    const context = createCodegenContext();
    const { push } = context;

    // 处理插值类型中的"const { toDisplayString: _toDisplayString } = Vue"部分
    // const VueBegining = "Vue";
    // // const helpers = ["toDisplayString"];
    // const aliasHelper = (s) => `${s} : _${s}`;
    // // push(`const { ${helpers.map(aliasHelper).join(", ")} } = ${VueBegining}`);

    // // 优化
    // push(`const { ${ast.helpers.map(aliasHelper).join(", ")} } = ${VueBegining}`);
    // push("\n");

    // 进行抽离
    getFunctionPreamble(ast, context);



    // 处理字符串类型
    const functionName = "render";
    const args = ["_ctx", "_cache"];
    const argsSignature = args.join(", ");
    
    push(` function ${functionName}(${argsSignature}){`);
    push(`return `);

    genNode(ast.codegenNode, context);
    push("}");

    return {
        code : context.code,
    }

}

function getFunctionPreamble(ast, context) {
    const { push, helper } = context;
    const VueBegining = "Vue";
    // const helpers = ["toDisplayString"];
    const aliasHelper = (s) => `${helper(s)} : _${helper(s)}`;
    // push(`const { ${helpers.map(aliasHelper).join(", ")} } = ${VueBegining}`);

    // 优化
    if (ast.helpers.length > 0) {
        // console.log("🚀 ~ file: codegen.ts:102 ~ getFunctionPreamble ~ ast.helpers:", ast.helpers)
        push(`const { ${ast.helpers.map(aliasHelper).join(", ")} } = ${VueBegining}`);
    }

    push("\n");
    push(`return`);
}

function createCodegenContext() : any {
    const context = {
        code: "",
        // 实现code字符串的拼接
        push(source) {
            context.code += source;
        },
        helper(key) {
            return `${helperMapName[key]}`;
        }
    }

    return context;
}

function genNode(node, context) {
    switch (node.type) {
        // 处理文本类型
        case NodeTypes.TEXT:
            genText(node, context);
            break;
        // 处理插值类型
        case NodeTypes.INTERPOLATION:
            genInterpolation(node, context);
            break;
        // 处理插值中的简单表达式
        case NodeTypes.SIMPLE_EXPRESSION:
            genExpression(node, context);
            break;
        // 处理元素类型
        case NodeTypes.ELEMENT:
            genElement(node, context);
            break;
        // 处理联合类型
        case NodeTypes.COMPOUND_EXPRESSION:
            genCompundExpression(node, context);
            break;
        default:
            break;
            
    }
}

function genCompundExpression(node, context) {
    const { push } = context;
    const children = node.children;
    for (let i = 0; i < children.length; i++) {
        const child = children[i];
        if (isString(child)) {
            push(child);
        } else {
            genNode(child, context);
        }
    }
}


function genElement(node, context) {
    const { push, helper } = context;
    const { tag, children, props } = node;
    push("(");

    console.log("node", node.children)
    // 纯元素类型（例如<div></div>）
    // 实现元素类型中的"(_openBlock(), _createElementBlock("div"))"部分
    if ((children && isEmpty(children)) || !children) {
        push(`_${helper(OPENBLOCK)}(), _${helper(CREATE_ELEMENT_BLOCK)}("${tag}"`)
    } else {
        // 联合类型
        // 实现联合类型中的"(_openBlock(), _createElementBlock("div", null, "hi" + _toDisplayString(_ctx.message), 1 /* TEXT */))"部分
        push(`_${helper(OPENBLOCK)}(), _${helper(CREATE_ELEMENT_BLOCK)}(`);
        genNodeList(genNullable([tag, props, children]), context);
        if (node.children) {
            
            if (node.children.type === NodeTypes.COMPOUND_EXPRESSION) {
                const compundChildren = node.children.children;
                for (let i = 0; i < compundChildren.length; i++) {
                    const child = compundChildren[i];
                    if (typeof child  !== "object") {
                        continue;
                    }
                    
                    if (child.type === NodeTypes.INTERPOLATION) {
                        push(`, ${child.content && child.content.type} /* TEXT */`)
                    }
                }
            }
        }
    }
    
    push(")");
    push(")");
}

function genNodeList(nodes, context) {
  const { push } = context;

  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i];
    if (isString(node)) {
      node === "null" ? push(null) : push(`"${node}"`);
    } else {
      genNode(node, context);
    }

    if(i < nodes.length -1){
      push(", ")
    }
  }
}

function genNullable(args: any) {
  return args.map((arg) => arg || "null");
}

function genExpression(node, context) {
    const { push } = context;
    // push(`_ctx.${node.content}`);
    // 將添加"_ctx."的逻辑抽离至transformExpression函数中
    push(`${node.content}`);
}

function genText(node, context) {
    const { push } = context;
    push(`'${node.content}'`);
}


// 生成插值类型中的"_toDisplayString(_ctx.message)"部分
function genInterpolation(node, context) {
    const { push, helper } = context;
    // push(`_toDisplayString(`);

    // 优化：使用helper()进行TO_DISPLAY_STRING和字符串"toDisplayString"的映射和"_"的拼接，避免直接书写字符串
    push(`_${helper(TO_DISPLAY_STRING)}(`);
    // 进入genNode中的处理插值中的简单表达式分支
    genNode(node.content, context);
    push(")");
}