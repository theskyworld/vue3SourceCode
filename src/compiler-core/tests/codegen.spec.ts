import { transform } from "../src/transform";
import { generate } from "../src/codegen";
import { baseParse } from "../src/parse"
import { transformExpression } from "../src/utils/transform/transformExpression";
import { transformElement } from "../src/utils/transform/transformElement";
import { transformUnion } from "../src/utils/transform/transformUnion";
import { transformText } from "../src/utils/transform/transformText";


describe("codegen", () => {
    it("string", () => {
        // 基于文本节点，内容为"hi"来生成一个render函数
        const ast = baseParse('hi');
        transform(ast);
        const { code } = generate(ast);
        

        // 生成code的快照
        // 生成的文件在_snapshots_文件夹中查看
        // 每次code更新之后要执行npm test -- -u 来进行相应快照的更新
        // 否则将导致测试失败
        expect(code).toMatchSnapshot();
    });

    
    
    it("interpolation", () => {
        // 基于插值，内容为message来生成一个render函数
        const ast = baseParse('{{message}}');
        transform(ast, {
            nodeTransforms: [transformExpression],
        });
        const { code } = generate(ast);
        

        // 生成code的快照
        // 生成的文件在_snapshots_文件夹中查看
        // 每次code更新之后要执行npm test -- -u 来进行快照的更新
        // 否则将导致测试失败
        expect(code).toMatchSnapshot();
    });

    it("element", () => {
        // 基于元素类型来生成一个render函数
        const ast = baseParse('<div></div>');
        transform(ast, {
            nodeTransforms: [transformElement],
        });
        const { code } = generate(ast);
        

        // 生成code的快照
        // 生成的文件在_snapshots_文件夹中查看
        // 每次code更新之后要执行npm test -- -u 来进行快照的更新
        // 否则将导致测试失败
        expect(code).toMatchSnapshot();
    });


    it("union", () => {
        // 基于联合类型来生成一个render函数
        const ast = baseParse('<div>hi,{{message}}</div>');
        transform(ast, {
            nodeTransforms : [transformExpression, transformElement, transformText], 
        });
        const { code } = generate(ast);
        

        // 生成code的快照
        // 生成的文件在_snapshots_文件夹中查看
        // 每次code更新之后要执行npm test -- -u 来进行快照的更新
        // 否则将导致测试失败
        expect(code).toMatchSnapshot();
    })
})