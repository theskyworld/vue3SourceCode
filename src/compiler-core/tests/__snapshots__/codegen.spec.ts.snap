// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen element 1`] = `
"const { createElementBlock : _createElementBlock, openBlock : _openBlock } = Vue
return function render(_ctx, _cache){return (_openBlock(), _createElementBlock("div"))}"
`;

exports[`codegen interpolation 1`] = `
"const { toDisplayString : _toDisplayString } = Vue
return function render(_ctx, _cache){return _toDisplayString(_ctx.message)}"
`;

exports[`codegen string 1`] = `
"
return function render(_ctx, _cache){return 'hi'}"
`;

exports[`codegen union 1`] = `
"const { toDisplayString : _toDisplayString, createElementBlock : _createElementBlock, openBlock : _openBlock } = Vue
return function render(_ctx, _cache){return (_openBlock(), _createElementBlock("div", null, 'hi,' + _toDisplayString(_ctx.message), 1 /* TEXT */))}"
`;
